driver-sync/src/test/functional/com/mongodb/client/FailPoint.java

```diff
@@ -15,33 +15,95 @@
  */
 package com.mongodb.client;
 
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoCredential;
+import com.mongodb.ServerAddress;
+import com.mongodb.ServerApi;
+import com.mongodb.connection.ClusterConnectionMode;
+import com.mongodb.connection.ClusterSettings;
+import com.mongodb.connection.ClusterType;
+import com.mongodb.connection.ConnectionPoolSettings;
 import org.bson.BsonDocument;
 import org.bson.BsonString;
 import org.bson.conversions.Bson;
 
+import java.util.Collections;
+
 public final class FailPoint implements AutoCloseable {
     private final BsonDocument failPointDocument;
     private final MongoClient client;
+    private final boolean close;
 
-    private FailPoint(final BsonDocument failPointDocument, final MongoClient client) {
+    private FailPoint(final BsonDocument failPointDocument, final MongoClient client, final boolean close) {
         this.failPointDocument = failPointDocument.toBsonDocument();
         this.client = client;
+        this.close = close;
     }
 
     /**
-     * @param configureFailPointDocument A document representing {@code configureFailPoint} command to be issued as is via
+     * @param configureFailPointDoc A document representing {@code configureFailPoint} command to be issued as is via
      * {@link com.mongodb.client.MongoDatabase#runCommand(Bson)}.
      */
-    public static FailPoint enable(final BsonDocument configureFailPointDocument, final MongoClient client) {
-        FailPoint result = new FailPoint(configureFailPointDocument, client);
-        client.getDatabase("admin").runCommand(configureFailPointDocument);
+    public static FailPoint enable(final BsonDocument configureFailPointDoc, final MongoClientSettings clientSettingsTemplate,
+            final ServerAddress serverAddress) {
+        MongoClientSettings.Builder clientSettingsBuilder = MongoClientSettings.builder()
```

Comment:
This code does not clear listeners, but they all seem to be erasable in principle, and the approach seems to be possible. In fact, the approach with `MongoClientSettings.builder(clientSettingsTemplate)` is the first one I tried, but I was unable to make the resulting settings work in all Evergreen environments (almost certainly because of my mistakes), so I changed the approach. However, I may give it another try.

Reply:
Maybe we could instead arrange for the clientSettingsTemplate to _not_ have listeners, instead of trying to remove them here.  I'm not sure why it does have them in the first place.  Or are you saying: in case it has them, they need to be removed?

Reply:
> Or are you saying: in case it has them, they need to be removed?

Exactly

Reply:
It's pretty rare for listeners to be added.  For test code like this I think it's fine to just say that it's up to the caller to provide a client settings template that doesn't contain listeners unless for some reason they are desired.

Reply:
Refactored to use `MongoClientSettings.builder(clientSettingsTemplate)`, and the listeners are also cleared.